File: backend/pop3/src/main/java/com/mail/ann/backend/pop3/CommandSetFlag.java
package com.mail.ann.backend.pop3;


import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.mail.ann.mail.Flag;
import com.mail.ann.mail.MessagingException;
import com.mail.ann.mail.store.pop3.Pop3Folder;
import com.mail.ann.mail.store.pop3.Pop3Message;
import com.mail.ann.mail.store.pop3.Pop3Store;
import org.jetbrains.annotations.NotNull;


class CommandSetFlag {
    private final Pop3Store pop3Store;


    CommandSetFlag(Pop3Store pop3Store) {
        this.pop3Store = pop3Store;
    }

    void setFlag(@NotNull String folderServerId, @NotNull List<String> messageServerIds, @NotNull Flag flag,
            boolean newState) throws MessagingException {

        Pop3Folder remoteFolder = pop3Store.getFolder(folderServerId);
        if (!remoteFolder.isFlagSupported(flag)) {
            return;
        }

        try {
            remoteFolder.open();
            List<Pop3Message> messages = new ArrayList<>();
            for (String uid : messageServerIds) {
                messages.add(remoteFolder.getMessage(uid));
            }

            if (messages.isEmpty()) {
                return;
            }
            remoteFolder.setFlags(messages, Collections.singleton(flag), newState);
        } finally {
            remoteFolder.close();
        }
    }
}

File: backend/imap/src/main/java/com/mail/ann/backend/imap/CommandSetFlag.kt
package com.mail.ann.backend.imap

import com.mail.ann.mail.Flag
import com.mail.ann.mail.store.imap.ImapStore
import com.mail.ann.mail.store.imap.OpenMode

internal class CommandSetFlag(private val imapStore: ImapStore) {

    fun setFlag(folderServerId: String, messageServerIds: List<String>, flag: Flag, newState: Boolean) {
        if (messageServerIds.isEmpty()) return

        val remoteFolder = imapStore.getFolder(folderServerId)
        if (!remoteFolder.exists()) return

        try {
            remoteFolder.open(OpenMode.READ_WRITE)
            if (remoteFolder.mode != OpenMode.READ_WRITE) return

            val messages = messageServerIds.map { uid -> remoteFolder.getMessage(uid) }

            remoteFolder.setFlags(messages, setOf(flag), newState)
        } finally {
            remoteFolder.close()
        }
    }
}
